plugins/auth/fps_auth/backends.py

```diff
@@ -0,0 +1,104 @@
+from typing import Optional
+
+import httpx
+from fastapi import Depends, status
+from fastapi.security.base import SecurityBase
+from fastapi_users.authentication import BaseAuthentication, CookieAuthentication  # type: ignore
+from fastapi_users import FastAPIUsers, BaseUserManager  # type: ignore
+from starlette.requests import Request
+
+from .models import (
+    User,
+    UserCreate,
+    UserUpdate,
+    UserDB,
+)
+from .config import get_auth_config
+from .db import secret, get_user_db
+
+
+class NoAuth(SecurityBase):
+    def __call__(self):
+        return
+
+
+class NoAuthAuthentication(BaseAuthentication):
+    def __init__(self, user: UserDB, name: str = "noauth"):
+        super().__init__(name, logout=False)
+        self.user = user
+        self.scheme = NoAuth()
+
+    async def __call__(self, credentials, user_manager):
+        # always return the user no matter what
+        return self.user
```

Comment:
What I can see is that when jupyverse is launched with `--authenticator.mode=noauth`, authentication fails because this is never called.

Reply:
Ok, I know where this comes from; this is a bit nasty. In FastAPI, security schemes are dependencies responsible for returning credentials from a request (say, a token in headers, a cookie, or whatever).

In our case, we defined a custom scheme as this:

```py
class NoAuth(SecurityBase):
    def __call__(self):
        return
```

It always returns `None` no matter the request. It makes sense, we don't actually care about the value here. However, in FastAPI Users, we do [this](https://github.com/fastapi-users/fastapi-users/blob/0826ec529915e80c217f1ddc52f76d1c1c71d223/fastapi_users/authentication/__init__.py#L141-L147):

```py
        for backend in self.backends:
            if backend in enabled_backends:
                token: str = kwargs[name_to_variable_name(backend.name)]
                if token:
                    user = await backend(token, user_manager)
                    if user:
                        break
```

So, if the token is `None`, we pass and never actually call the backend. I can't change this in FastAPI Users because this is necessary to know when no token is retrieved from the request.

So, the ugly fix could be this:

```py
class NoAuth(SecurityBase):
    def __call__(self):
        return "noauth"  # or any value you want different from None
```

Not ideal, but this is the best I can imagine for now 😅

Reply:
Works for me, thanks a lot!

Reply:
I see! Well, what you could do is to retrieve it dynamically then inside `__call__` by making a query to the database. You could use a static e-mail for user or add a flag to the model `is_noauth_user` to retrieve it easily.